lib/test.js

```diff
@@ -699,10 +724,19 @@ class Test {
 		};
 	}
 
-	finishPromised() {
-		return new Promise(resolve => {
+	async finishPromised() {
+		const testResult = await new Promise(resolve => {
 			resolve(this.finish());
 		});
+
+		const teardownResult = await this.runTeardowns();
```

Comment:
We should run the teardown right after `verifyAssertions()`: https://github.com/avajs/ava/blob/276c16afc85bddc2c68592a002a96eb4cf3521cc/lib/test.js#L672

Then combined with calling `saveFirstError()` if the teardown crashes, the rest of the function will correctly record the test as having failed.

What's interesting is that the teardown function may still run assertions. The test would fail if your only assertion is run inside a teardown, which is fine. And they wouldn't count towards the plan. But we should document that, and test for it.

Reply:
This was the part I was most unsure of: when to actually run them. Sounds good.

Yeah, I know. It's a little weird, but that's how it is, I guess. Will do, thanks!

Reply:
Hum, adding a test for _running assertion in `teardown()`_ revealed: 

1. `Error: Assertion <X>, but test has already finished`
2. The `assertCount` is indeed incremented

What's your take on that?

Reply:
For example, see: https://github.com/avajs/ava/blob/da09b00eee34ed7304a7fb11072979cc4e29439f/lib/test.js#L324